Enhance MASTG-TEST-0355, MASTG-TEST-0356, MASTG-TEST-0357 with Example Attack Scenario and Further Validation (#3850)
OWASP/mastg: tests-beta/android/MASVS-PLATFORM/MASTG-TEST-0356.md

## Overview

> If an app exports a content provider without requiring permissions, any app on the device can directly query its underlying database using `ContentResolver` or the `adb shell content` command. Even when a permission is declared, a misconfigured protection level (for example, `android:protectionLevel="normal"`) allows any requesting app to obtain it automatically, effectively bypassing the restriction. This test verifies at runtime whether the app's exported content providers are accessible without the required permissions.

Reviewer: ```suggestion
If an app exports a content provider without requiring permissions, any app on the device can directly query its underlying database using [`ContentResolver`](https://developer.android.com/reference/android/content/ContentResolver) or using the `adb shell content` command. Even when a permission is declared, a misconfigured protection level (for example, `android:protectionLevel="normal"`) allows any requesting app to obtain it automatically, effectively bypassing the restriction. This test verifies at runtime whether the app's exported content providers are accessible without the required permissions.
```

**Example Attack Scenario:**

> Suppose a finance app stores transaction records in a database exposed through an exported content provider with no declared `android:readPermission`.

Reviewer: ```suggestion
Suppose a health app stores medical records in a database exposed through an exported content provider with no declared `android:readPermission` and protection level.
```

1. An attacker identifies the exported provider's authority from the AndroidManifest.

> 2. The attacker uses `adb shell content query` to query the provider's URI without any credentials.

Reviewer: ```suggestion
2. The attacker uses `adb shell content query` to query the provider's URI without any restrictions.
```

3. The content provider returns all database rows without checking the caller's identity.

> 4. The attacker reads account numbers, balances, and transaction history directly from the victim's device.

Reviewer: ```suggestion
4. The attacker reads PII or medical information directly from the app.
5. With this knowledge the attacker crafts a malicious app by using `ContentResolver` and trying to lure potential victims into side-loading the app on their device. 
6. When a victim is installing the malicious app it will query for the data and sent it to the attackers server.```

## Overview

If an app exports a content provider without requiring permissions, any app on the device can directly query its underlying database using [`ContentResolver`](https://developer.android.com/reference/android/content/ContentResolver) or using the `adb shell content` command. Even when a permission is declared, a misconfigured protection level (for example, `android:protectionLevel="normal"`) allows any requesting app to obtain it automatically, effectively bypassing the restriction. This test verifies at runtime whether the app's exported content providers are accessible without the required permissions.

**Example Attack Scenario:**

Suppose a health app stores medical records in a database exposed through an exported content provider with no declared `android:readPermission` and protection level.

1. An attacker identifies the exported provider's authority from the AndroidManifest.
2. The attacker uses `adb shell content query` to query the provider's URI without any restrictions.
3. The content provider returns all database rows without checking the caller's identity.
4. The attacker reads PII or medical information directly from the app.

> 5. With this knowledge the attacker crafts a malicious app by using `ContentResolver` and trying to lure potential victims into side-loading the app on their device. 

Author: ```suggestion
5. With this knowledge the attacker crafts a malicious app by using `ContentResolver` and trying to lure potential victims into side-loading the app on their device.
```
